teal/packer/plugin.tl

```diff
@@ -0,0 +1,238 @@
+end
+
+local function normcond(x: string|{string}): {string}
+  if x is string then
+    return { x }
+  end
+  return x as {string}
+end
+
+local function normkeys(x: string|{{string,string}}): {{string,string}}
+  if x is string then
+    return {{ '', x as string }}
+  end
+  return x as {{string,string}}
+end
+
+local function normrun(x: string|function|{string|function}): {string|function}
+  if x is function or x is string then
+    return { x }
+  end
+  return x as {string|function}
+end
+
+--- The main logic for adding a plugin (and any dependencies) to the managed set
+-- Can be invoked with (1) a single plugin spec as a string, (2) a single plugin spec table, or (3)
+-- a list of plugin specs
+local function process_spec(
+  spec0: string|M.UserSpec,
+  plugins: {string:M.Plugin},
+  required_by: M.Plugin
+): {string:M.Plugin}
+  local spec = normspec(spec0)
+
+  if #spec > 1 then
+    local r: {string:M.Plugin} = {}
+    for _, s in ipairs(spec) do
+      r = vim.tbl_extend('error', r, process_spec(s, plugins, required_by))
+    end
+    return r
+  end
+
+  local id = spec[1] as string
+  spec[1] = nil
+
+  if id == nil then
+    log.warn('No plugin name provided!')
+    return {}
+  end
+
+  local name, path = get_plugin_name(id)
+
+  if name == '' then
+    log.warn(fmt('"%s" is an invalid plugin name!', id))
+    return {}
+  end
+
+  if plugins[name] then
+    if required_by then
+      plugins[name].required_by = plugins[name].required_by or {}
+      table.insert(plugins[name].required_by, required_by.name)
+    else
+      log.warn(fmt('Plugin "%s" is specified more than once!', name))
+    end
+
+    return {[name] = plugins[name]}
+  end
+
+  local url, ptype = guess_plugin_type(path)
+
+  local plugin: M.Plugin = {
+    name        = name,
+    full_name   = get_plugin_full_name(name, spec),
+    branch      = spec.branch,
+    rev         = spec.rev,
+    tag         = spec.tag,
+    commit      = spec.commit,
+    keys        = normkeys(spec.keys),
+    event       = normcond(spec.event),
+    ft          = normcond(spec.ft   ),
+    cmd         = normcond(spec.cmd  ),
+    cond        = spec.cond,
+    run         = normrun(spec.run),
+    lock        = spec.lock,
+    url         = remove_ending_git_url(url),
+    type        = ptype,
+    config      = spec.config,
+    required_by = required_by and {required_by.name} or nil,
+    revs        = {}
+  }
+
+  plugins[name] = plugin
+
+  if plugin.opt == nil then
+    plugin.opt = plugin.keys ~= nil
+      or plugin.ft ~= nil
+      or plugin.cmd ~= nil
+      or plugin.event ~= nil
+      or plugin.cond ~= nil
+      or (required_by or {}).opt
+  end
+
+  plugin.install_path = util.join_paths(plugin.opt and config.opt_dir or config.start_dir, name)
+  plugin.loaded = not plugin.opt and vim.loop.fs_stat(plugin.install_path) ~= nil
+
+  if spec.requires then
+    if required_by then
+      log.warn(fmt('(%s) Nested requires are not support', name))
```

Comment:
Yeah we could go either way. I'm on the side that nesting here is bad practice, and we should just enforce a flat format. Recursing is possible, but I think it might introduce some edge cases that Teal isn't quite good enough to cover us for.

I'm also thinking about making `requires` just a list of strings and if the user wants to customise it, then they need to add it to the spec at the top level. That would mean each plugin would have a single source of truth, rather than worrying about having to merge specs.

 E.g:

```lua
packer.startup{{
    {'a', requires = {'b'}},
    {'b', branch = 'mybranch'},  -- Customise branch of 'b'
}}
```

Otherwise handling the following is very difficult:

```lua
packer.startup{{
    {'a', requires = { {'b', branch = 'otherbranch'}}},
    {'b', branch = 'mybranch'},
    -- need to detect consistency of `branch` in `b` and raise an error in this case. 
}}
```

Reply:
The point of the nesting is that if you remove all requiring plugins, this also removes the required plugin. I don't see how this is possible with a flat format?

Reply:
It isn't, that's one of the downsides (a minor one IMO). Otherwise, we need to do this consistency checking recursively, which is pretty complicated and avoidable.

To note:

```lua
packer.startup{{
    {'a', requires = { 'b' }}
}}
```

will still install `b`, and remove it when it is removed from `requires`, you just can't nest any more than that. I think that cpatures 95% of usecases.

Reply:
Agreed, I just wanted to mention this for the sake of completeness. 

Although I do have
```lua
  {
    'nvim-telescope/telescope.nvim',
    requires = {
      'nvim-lua/plenary.nvim',
      { 'nvim-telescope/telescope-fzf-native.nvim', run = 'make' },
      { 'nvim-telescope/telescope-frecency.nvim', requires = 'kkharji/sqlite.lua' },
    },
  },
```
and I wager `telescope-frecency` is not so niche ;) (Or at least it shouldn't be, as it's rather a game-changer.)

Reply:
It's decent usecase. I'll keep it in mind.

Reply:
Good plan. For now I might just add it as a todo.